src/main/java/com/mongodb/kafka/connect/source/MongoCopyDataManager.java

```diff
@@ -0,0 +1,144 @@
+ * <li>Each copy existing task, runs an aggregation pipeline on the namespace, to mimic an insert document and adds the results to the
+ *     queue.</li>
+ * <li>The poll method returns documents from the queue.</li>
+ * </ol>
+ */
+class MongoCopyDataManager implements AutoCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoCopyDataManager.class);
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final AtomicInteger namespacesToCopy;
+    private final MongoSourceConfig sourceConfig;
+    private final MongoClient mongoClient;
+    private final ExecutorService executor;
+    private final ConcurrentLinkedQueue<BsonDocument> queue = new ConcurrentLinkedQueue<>();
+
+    MongoCopyDataManager(final MongoSourceConfig sourceConfig, final MongoClient mongoClient) {
+        this.sourceConfig = sourceConfig;
+        this.mongoClient = mongoClient;
+
+        String database = sourceConfig.getString(DATABASE_CONFIG);
+        String collection = sourceConfig.getString(COLLECTION_CONFIG);
+
+        List<MongoNamespace> namespaces;
+        if (database.isEmpty()) {
+            namespaces = getCollections(mongoClient);
+        } else if (collection.isEmpty()) {
+            namespaces = getCollections(mongoClient, database);
+        } else {
+            namespaces = singletonList(createNamespace(database, collection));
+        }
+        LOGGER.info("Copying existing data on the following namespaces: {}", namespaces);
+        namespacesToCopy = new AtomicInteger(namespaces.size());
+        executor = Executors.newFixedThreadPool(
+                Math.min(namespacesToCopy.get(), sourceConfig.getInt(COPY_EXISTING_MAX_THREADS_CONFIG)));
+        namespaces.forEach(n -> executor.submit(() -> copyDataFrom(n)));
+    }
+
+    Optional<BsonDocument> poll() {
+        if (namespacesToCopy.get() == 0) {
+            close();
+        }
+        return Optional.ofNullable(queue.poll());
+    }
+
+    boolean isCopying() {
+        return namespacesToCopy.get() > 0 || !queue.isEmpty();
+    }
+
+    @Override
+    public void close() {
+        if (!closed.getAndSet(true)) {
+            LOGGER.debug("Shutting down executors");
+            executor.shutdownNow();
+        }
+    }
+
+    private void copyDataFrom(final MongoNamespace namespace) {
+        LOGGER.debug("Copying existing data from: {}", namespace.getFullName());
+        mongoClient.getDatabase(namespace.getDatabaseName())
+                .getCollection(namespace.getCollectionName(), BsonDocument.class)
+                .aggregate(createPipeline(namespace))
+                .forEach((Consumer<? super BsonDocument>) queue::add);
+        namespacesToCopy.decrementAndGet();
+    }
+
+    private List<Bson> createPipeline(final MongoNamespace namespace) {
+        List<Bson> pipeline = new ArrayList<>();
+        pipeline.add(BsonDocument.parse("{$replaceRoot: {newRoot: {"
+                + "_id: {_id: '$_id', copyingData: true}, "
```

Comment:
the `copyingData` flag is used by the MongoSourceTask to understand if the resume token is from the copying data process.